arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/raise/RaiseAccumulate.kt

```diff
@@ -476,6 +478,45 @@ public inline fun <Error, A, B> Raise<NonEmptyList<Error>>.mapOrAccumulate(
   return error.toNonEmptyListOrNull()?.let { raise(it) } ?: results
 }
 
+/**
+ * Accumulate the errors obtained by executing the [transform] over every element of [NonEmptyList].
+ */
+@RaiseDSL
+public inline fun <Error, A, B> Raise<NonEmptyList<Error>>.mapOrAccumulate(
+  nonEmptyList: NonEmptyList<A>,
+  @BuilderInference transform: RaiseAccumulate<Error>.(A) -> B
+): NonEmptyList<B> {
+  val error = mutableListOf<Error>()
+  val results = ArrayList<B>(nonEmptyList.collectionSizeOrDefault(10))
+  for (item in nonEmptyList) {
+    fold<NonEmptyList<Error>, B, Unit>(
+      { transform(RaiseAccumulate(this), item) },
+      { errors -> error.addAll(errors) },
+      { results.add(it) }
+    )
+  }
+  return error.toNonEmptyListOrNull()?.let { raise(it) } ?: requireNotNull(results.toNonEmptyListOrNull())
+}
+
+/**
+ * Accumulate the errors obtained by executing the [transform] over every element of [NonEmptySet].
+ */
+@RaiseDSL
+public inline fun <Error, A, B> Raise<NonEmptyList<Error>>.mapOrAccumulate(
+  nonEmptySet: NonEmptySet<A>,
+  @BuilderInference transform: RaiseAccumulate<Error>.(A) -> B
+): NonEmptySet<B> {
+  val error = mutableListOf<Error>()
+  val results = ArrayList<B>(nonEmptySet.collectionSizeOrDefault(10))
+  for (item in nonEmptySet) {
+    fold<NonEmptyList<Error>, B, Unit>(
+      { transform(RaiseAccumulate(this), item) },
+      { errors -> error.addAll(errors) },
+      { results.add(it) }
+    )
+  }
+  return error.toNonEmptyListOrNull()?.let { raise(it) } ?: requireNotNull(results.toNonEmptySetOrNull())
```

Comment:
For `NonEmptySet` we can provide a more optimised version:

```kotlin
@RaiseDSL
public inline fun <Error, A, B> Raise<NonEmptyList<Error>>.mapOrAccumulate(
  iterable: NonEmptySet<A>,
  @BuilderInference transform: RaiseAccumulate<Error>.(A) -> B
): NonEmptySet<B> {
  val error = mutableListOf<Error>()
  val results = HashSet<B>(iterable.collectionSizeOrDefault(10))
  for (item in iterable) {
    fold<NonEmptyList<Error>, B, Unit>(
      { transform(RaiseAccumulate(this), item) },
      { errors -> error.addAll(errors) },
      { results.add(it) }
    )
  }
  return error.toNonEmptyListOrNull()?.let { raise(it) } ?: requireNotNull(results.toNonEmptySetOrNull())
}
```

It requires improving the current `toNonEmptySetOrNull` APIs, they're currently defined over `Collection` instead of `Iterable` and no definitions available for `Set`.

```kotlin
public fun <T> Iterable<T>.toNonEmptySetOrNull(): NonEmptySet<T>? =
  firstOrNull()?.let { NonEmptySet(it, minus(it).toSet()) }

public fun <T> Iterable<T>.toNonEmptySetOrNone(): Option<NonEmptySet<T>> =
  toNonEmptySetOrNull().toOption()

public fun <T> Set<T>.toNonEmptySetOrNull(): NonEmptySet<T>? =
  firstOrNull()?.let { NonEmptySet(it, minus(it)) }

public fun <T> Set<T>.toNonEmptySetOrNone(): Option<NonEmptySet<T>> =
  toNonEmptySetOrNull().toOption()
```

Reply:
No, this is a newly added API so deprecation is not needed.